### PR TITLE
NNCF documentation update

### PR DESCRIPTION
**Pull request by AlexKoff88**
### Details:
 - Added NNCF description
 - Added QAT description
 - Added Filter pruning description
 - Added code snippets

### CODE REVIEW

**Comment by AlexKoff88** [comment]
cc'ed @alexsu52

File: docs/optimization_guide/nncf/code/pruning_tf.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            "algorithm": "quantization",  # 8-bit quantization with default settings
+        },
+    ]
+}
+nncf_config = NNCFConfig.from_dict(nncf_config_dict)
+nncf_config = register_default_init_args(nncf_config, train_dataset, batch_size=1) # train_dataset is an instance of tf.data.Dataset
+#! [nncf_congig]
+
+#! [wrap_model]
+model = KerasModel() # instance of the tensorflow.python.keras.models.Model
```
Comment (alexsu52) [inline docs/optimization_guide/nncf/code/pruning_tf.py:33]:
```suggestion
model = KerasModel() # instance of the tensorflow.keras.Model
```

File: docs/optimization_guide/nncf/code/qat_tf.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+#! [wrap_model]
+
+#! [distributed]
+compression_ctrl.distributed() # call it before the training
+#! [distributed]
+
+#! [tune_model]
+... # fine-tuning preparations, e.g. dataset, loss, optimizer setup, etc.
+# tune quantized model for 5 epochs the same way as the baseline
+model.fit(train_dataset, epochs=5)
```
Comment (alexsu52) [inline docs/optimization_guide/nncf/code/qat_tf.py:34]:
```suggestion
# create compression callbacks to control optimization parameters and dump compression statistics
compression_callbacks = create_compression_callbacks(compression_ctrl, log_dir="./compression_log")
# tune quantized model for 5 epochs the same way as the baseline
model.fit(train_dataset, epochs=5, callbacks=compression_callbacks)
```

File: docs/optimization_guide/nncf/code/pruning_tf.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+#! [wrap_model]
+
+#! [distributed]
+compression_ctrl.distributed() # call it before the training
+#! [distributed]
+
+#! [tune_model]
+... # fine-tuning preparations, e.g. dataset, loss, optimizer setup, etc.
+# tune quantized model for 50 epochs as the baseline
+model.fit(train_dataset, epochs=50)
```
Comment (alexsu52) [inline docs/optimization_guide/nncf/code/pruning_tf.py:44]:
```suggestion
# create compression callbacks to control pruning parameters and dump compression statistics
compression_callbacks = create_compression_callbacks(compression_ctrl, log_dir="./compression_log")
# tune quantized model for 50 epochs as the baseline
model.fit(train_dataset, epochs=50, callbacks=compression_callbacks)
```

File: docs/optimization_guide/nncf/code/pruning_tf.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+#! [tune_model]
+
+#! [export]
+compression_ctrl.export_model("compressed_model.pb", save_format='frozen_graph') #export to Frozen Graph
```
Comment (alexsu52) [inline docs/optimization_guide/nncf/code/pruning_tf.py:48]:
```suggestion
compression_ctrl.export_model("compressed_model.pb") #export to Frozen Graph
```

File: docs/optimization_guide/nncf/code/pruning_tf.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+#! [export] 
+
+#! [save_checkpoint]
+from nncf.tensorflow.utils.state import TFCompressionState
```
Comment (alexsu52) [inline docs/optimization_guide/nncf/code/pruning_tf.py:52]:
```suggestion
from nncf.tensorflow.utils.state import TFCompressionState 
from nncf.tensorflow.callbacks.checkpoint_callback import CheckpointManagerCallback
```

File: docs/optimization_guide/nncf/filter_pruning.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+@sphinxtab{TensorFlow 2}
+
+@snippet docs/optimization_guide/nncf/code/pruning_tf.py tune_model
+
+@endsphinxtab
+
+@endsphinxtabset
+
+> **NOTE**: In the case of distributed multi-GPU training (not DataParallel), you should call `compression_ctrl.distributed()` before the fine-tuning that will inform optimization methods to do some adjustments to function in the distributed mode.
```
Comment (alexsu52) [inline docs/optimization_guide/nncf/filter_pruning.md:97]: same comment here as for QAT.

File: docs/optimization_guide/nncf/filter_pruning.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+@sphinxtabset
```
Comment (alexsu52) [inline docs/optimization_guide/nncf/filter_pruning.md:99]: ### 6. Multi-GPU distributed training?

File: docs/optimization_guide/nncf/introduction.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+![](../../img/nncf_workflow.png)
+
+## Optimization options
+There are two methods available to improve model performance with OpenVINO&trade;:
+- **Quantization-aware Training (QAT)** is a **recommended** method that provides fast and accurate results. Currently, a HW-compatible (CPU, GPU, VPU) QAT for 8-bit inference is available. For details, see [Quantization-aware Training](./qat.md) documentation.
+- **Filter Pruning** is used to remove unnecessary or redundant filters from Convolutional Neural Networks. It is usually not used stand-alone but can be stacked with QAT to get additional speedup on top of it. This method is resource-demanding and requires model retraining which can take a significant amount of time in the general case. Thus, we recommend trying it when 8-bit quantization does not provide enough performance improvement. For details, see [Filter Pruning](./filter_pruning.md) documentation.
+
+## Installation
+NNCF is open-sourced on [GitHub](https://github.com/openvinotoolkit/nncf) and distributed as a separate package. It is also available on PyPI. We recommend installing it to the Python* environment where the framework is installed.
+
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply docs/optimization_guide/nncf/introduction.md:33]: The we should mention ONNX and probably IR in the future which will sound wordy )

[the remark this answers is not in the document]
Comment (AlexKoff88) [reply docs/optimization_guide/nncf/introduction.md:33]: I would think about whether should we guide the user to install NNCF into the same environment or create a separate one? The separate could be preferable since the right way is to create a separate script for model optimization (based on the original training script, however).

File: docs/optimization_guide/nncf/introduction.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+### Install from PyPI
+To install the latest released version via pip manager run the following command:
+```
+pip install nncf
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply docs/optimization_guide/nncf/introduction.md:37]: I presume that the environment already has a source framework installed. @alexsu52, what is your recommendation here?

File: docs/optimization_guide/nncf/qat.md [hunk 10]
```diff
@@ -0,0 +1,173 @@
+# Quantization-aware Training (QAT) {#qat_introduction}
+
+## Introduction
+Quantization-aware Training is a popular method that allows quantizing a model and applying fine-tuning to restore accuracy degradation caused by quantization. In fact, this is the most accurate quantization method. This document describes how to apply QAT from the Neural Network Compression Framework (NNCF) to get 8-bit quantized models. This assumes that you are knowledgeable in Python* programming and familiar with the training code for the model in the source DL framework.
+
+## Using NNCF QAT
+Here, we provide the steps that are required to integrate QAT from NNCF into the training script written with PyTorch or TensorFlow 2:
+
+> **NOTE**: Currently, NNCF for TensorFlow 2 supports optimization of the models created using Keras [Sequesntial API](https://www.tensorflow.org/guide/keras/sequential_model).
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply docs/optimization_guide/nncf/qat.md:9]: Fixed

File: docs/optimization_guide/nncf/code/qat_tf.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    "compression": {
+        "algorithm": "quantization",  # 8-bit quantization with default settings
+    },
+}
+nncf_config = NNCFConfig.from_dict(nncf_config_dict)
+nncf_config = register_default_init_args(nncf_config, train_dataset, batch_size=1) # train_dataset is an instance of tf.data.Dataset
+#! [nncf_congig]
+
+#! [wrap_model]
+model = KerasModel() # instance of the tensorflow.python.keras.models.Model
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply docs/optimization_guide/nncf/code/qat_tf.py:23]: applied

File: docs/optimization_guide/nncf/code/pruning_tf.py (hunk 3 above)
Comment (AlexKoff88) [reply docs/optimization_guide/nncf/code/pruning_tf.py:44]: Applied

File: docs/optimization_guide/nncf/code/pruning_tf.py (hunk 4 above)
Comment (AlexKoff88) [reply docs/optimization_guide/nncf/code/pruning_tf.py:48]: Applied

File: docs/optimization_guide/nncf/code/pruning_tf.py (hunk 5 above)
Comment (AlexKoff88) [reply docs/optimization_guide/nncf/code/pruning_tf.py:52]: Applied

File: docs/optimization_guide/nncf/filter_pruning.md (hunk 7 above)
Comment (AlexKoff88) [reply docs/optimization_guide/nncf/filter_pruning.md:99]: I think the note is ok since it's really one additional line of code from the NNCF perspective. Or it's not?

Comment (AlexKoff88) [reply docs/optimization_guide/nncf/filter_pruning.md:118]: applied

File: docs/optimization_guide/nncf/introduction.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexsu52) [reply docs/optimization_guide/nncf/introduction.md:33]: That's a good question. As you mentioned, the separate env is preferable, but to be honest, I often install NNCF into the existing env to get optimization capabilities.

File: docs/optimization_guide/nncf/introduction.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (alexsu52) [reply docs/optimization_guide/nncf/introduction.md:37]: Can do like in the https://pypi.org/project/openvino-dev/

For example:
`To install with specific frameworks, use the pip install nncf[extras] command, where extras is a list of extras from the table below:`

**Comment by AlexKoff88** [comment]
@kblaszczak-intel, feel free to reassign to somebody more relevant.

**Comment by alexsu52** [comment]
@AlexKoff88, It looks like some comments are not resolved. Could you please double-check?

**Comment by AlexKoff88** [comment]
> @AlexKoff88, It looks like some comments are not resolved. Could you please double-check?

I realized and fixed it. Thanks!

File: docs/optimization_guide/nncf/code/pruning_tf.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
+# Copyright (C) 2018-2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+#! [imports]
+import tensorflow as tf
+
+from nncf import NNCFConfig
+from nncf.torch import create_compressed_model, register_default_init_args
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply docs/optimization_guide/nncf/code/pruning_tf.py:8]: Please, take a look at this comment.

File: docs/optimization_guide/nncf/code/pruning_tf.py (hunk 1 above)
Comment (alexsu52) [reply docs/optimization_guide/nncf/code/pruning_tf.py:33]: Please, take a look at this comment.

File: docs/optimization_guide/nncf/code/qat_tf.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
+# Copyright (C) 2018-2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+#! [imports]
+import tensorflow as tf
+
+from nncf import NNCFConfig
+from nncf.torch import create_compressed_model, register_default_init_args
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply docs/optimization_guide/nncf/code/qat_tf.py:8]: Please, take a look at this comment.
